### ISSUE TITLE
Feature Requested: Improve softmax and CrossEntropy backward efficiency

### ISSUE THREAD

**Issue by fufjvnvnf** [stevenygd/NN.jl#7]

**Comment by fufjvnvnf** [comment]
Attempted to use loop in old softmax cross entropy forward.
Old code:
```
    l.x = Y
    l.exp = exp(l.x)
    l.lsum = sum(l.exp,2)
    l.y = l.exp ./ l.lsum
```
Performance:
```
Old forward uses on average 0.0037759856240000013 seconds
...
Old forward allocates on average 370706.596 bytes
```
New code using loop:
```
    for i=1:m
      for j=1:n
        l.exp[i,j] = exp(l.x[i,j])
      end
      l.lsum[i] = sum(l.exp[i,:])
      for j=1:n
        l.y[i,j] = l.exp[i,j]/l.lsum[i]
        l.exp[i,j] = log(l.y[i,j])*label[i,j]
      end
      l.loss[i] = -sum(l.exp[i,:])
    end
```
Performance:
```
Old forward uses on average 0.017988343472000005 seconds
...
Old forward allocates on average 2.133216e6 bytes
```

Using loop in this scenario seems to only deteriorate performance,
